ludwig/features/image_feature.py

```diff
@@ -206,36 +218,39 @@ def _finalize_preprocessing_parameters(
                 )
         else:
             # User hasn't specified height and width.
-            # Default to first image, or infer from sample.
-            height, width = first_img_height, first_img_width
-
+            # Default to inferring from sample or first image.
             if preprocessing_parameters[INFER_IMAGE_DIMENSIONS]:
                 should_resize = True
-                sample_size = min(len(input_feature_col), preprocessing_parameters[INFER_IMAGE_SAMPLE_SIZE])
-                sample_images = [read_image(get_image_from_path(src_path, img))
-                                 for img in input_feature_col[:sample_size]]
-
-                if sample_images:
-                    height_avg = min(
-                        sum(x.shape[0] for x in sample_images) / len(sample_images),
-                        preprocessing_parameters[INFER_IMAGE_MAX_HEIGHT])
-                    width_avg = min(
-                        sum(x.shape[1] for x in sample_images) / len(sample_images),
-                        preprocessing_parameters[INFER_IMAGE_MAX_WIDTH])
 
-                    height, width = round(height_avg), round(width_avg)
-
-                    logger.debug("Inferring height: {0} and width: {1}".format(height, width))
-                else:
-                    logger.warning("Sample set for inference is empty, default to height and width of first image")
+                height_avg = min(
+                    sum(x.shape[0] for x in inferred_sample) / len(inferred_sample),
+                    preprocessing_parameters[INFER_IMAGE_MAX_HEIGHT])
+                width_avg = min(
+                    sum(x.shape[1] for x in inferred_sample) / len(inferred_sample),
+                    preprocessing_parameters[INFER_IMAGE_MAX_WIDTH])
+
+                height, width = round(height_avg), round(width_avg)
+                logger.debug("Inferring height: {0} and width: {1}".format(height, width))
+            elif first_image is not None:
```

Comment:
It looks like we can generalize this `first_image` code path as a special case of the `infer_image_dimensions` case. Specifically, what if we removed the `infer_image_dimensions` param and just used `infer_image_sample_size`. Then if the user sets this value to `1`, we get the same effect as setting `infew_image_dimensions=False`, right?

Reply:
sounds good!

Reply:
actually I just realized there is a slight difference -`infer_image_dimensions` also sets `should_resize = True`, which will prevent Ludwig from throwing an error on the first mismatch.

Reply:
I see. In this case, I would be in favor of making `should_resize=True` the default (because the default is now to infer image dimensions). Then if the user wants the old behavior, they would need to set `should_resize=False` manually. An additional thing we can do in this case, when `should_resize=False` and `infer_image_sample_size > 1` is raise the error if not all the elements of the sample match. Would that solve the issue?

Reply:
By user setting the older behavior, do you mean exposing `should_resize` and `user_specified_num_channels` through preprocessing parameters? Or meaning they would just edit `should_resize = False` in the code.

If we expose `should_resize` and `user_specified_num_channels` through preprocessing parameters, then those would just be replacing `infer_image_dimensions`, so I don't know if that would decrease complexity. There also could be conflicts (i.e. user specifies explicit width and height but then has `should_resize` set to False

Reply:
Hmm, yeah I see the issue here. I think given how much we would need to overhaul the config params to make it work, it's probably fine to leave it as-is for now.